src/target/riscv/riscv.c

```diff
@@ -1561,6 +1561,52 @@ int riscv_remove_watchpoint(struct target *target,
 	return ERROR_OK;
 }
 
+typedef enum {
+	M6_HIT_ERROR,
+	M6_HIT_NOT_SUPPORTED,
+	M6_NOT_HIT,
+	M6_HIT_BEFORE,
+	M6_HIT_AFTER,
+	M6_HIT_IMM_AFTER
+} mctrl6hitstatus;
+
+static mctrl6hitstatus check_mcontrol6_hit_status(struct target *target,
+		riscv_reg_t tdata1, uint64_t hit_mask)
+{
+	uint32_t hit0 = get_field(tdata1, CSR_MCONTROL6_HIT0);
+	uint32_t hit1 = get_field(tdata1, CSR_MCONTROL6_HIT1);
+	uint32_t hit_info = (hit1 << 1) | hit0;
+	if (hit_info == CSR_MCONTROL6_HIT0_BEFORE)
+		return M6_HIT_BEFORE;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_AFTER)
+		return M6_HIT_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_IMMEDIATELY_AFTER)
+		return M6_HIT_IMM_AFTER;
+
+	if (hit_info == CSR_MCONTROL6_HIT0_FALSE) {
+		riscv_reg_t tdata1_tests[] = {
+			set_field(tdata1, CSR_MCONTROL6_HIT0, 1),
+			set_field(tdata1, CSR_MCONTROL6_HIT1, 1),
+			set_field(tdata1, CSR_MCONTROL6_HIT0, 1) | field_value(CSR_MCONTROL6_HIT1, 1)
+		};
+		riscv_reg_t tdata1_test_rb;
+		for (uint64_t i = 0; i < ARRAY_SIZE(tdata1_tests); ++i) {
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_tests[i]) != ERROR_OK)
+				return M6_HIT_ERROR;
+			if (riscv_reg_get(target, &tdata1_test_rb, GDB_REGNO_TDATA1) != ERROR_OK)
+				return M6_HIT_ERROR;
+			if (riscv_reg_set(target, GDB_REGNO_TDATA1, tdata1_test_rb & ~hit_mask) != ERROR_OK)
+				return M6_HIT_ERROR;
+			if (tdata1_test_rb == tdata1_tests[i])
+				return M6_NOT_HIT;
+		}
+	}
```

Comment:
Once we learn for sure whether the `hit` bits are supported for a given trigger, we should cache that information in `struct riscv_info`, for example this way:

```c
yes_no_maybe_t mcontrol6_supports_hit[RISCV_MAX_TRIGGERS]; // initialized to YNM_MAYBE
```

This way the number of reads and writes of TDATA1 will be greatly reduced.

@sunnyzhu-learning if you decide to not implement this improvement in this merge request, please add at least a TODO comment into this function (and in that case we will also open an issue to track it).

Reply:
Add TODO, addressed.

Reply:
@JanMatCodasip, AFAIK the spec does not require for `hit` bit to be supported across all possible configurations of a trigger with a given type.
e.g. It is a Sdtrig-compliant system that supports `hit` field with `e` set (on breakpoints), but does not support `hit` field with `r` or `w` set (on watchpoints).
Therefore it is incorrect to cache support of `mcontrol6.hit` field just based on trigger index.

Reply:
@en-sc, I believe you're right, thanks for the corrections. It would be a bit extreme for hardware to support hit bits on one mcontrol6 configuration but not other, but still the current spec permits it. I have opened a proposal to address this in a future version of the spec: https://github.com/riscv/riscv-debug-spec/issues/1080

@sunnyzhu-learning Per Evgeniy's comment, the TODO should be removed. We cannot cache the information. I'm sorry for this extra back and forth.